File: androidSharedNative/src/main/java/io/github/lumkit/tweak/sharednative/RootFileService.java
package io.github.lumkit.tweak.sharednative;

import android.content.Intent;
import android.os.Bundle;
import android.os.IBinder;

import com.topjohnwu.superuser.ipc.RootService;

import java.nio.charset.StandardCharsets;
import java.util.Arrays;
import java.util.Collections;

public final class RootFileService extends RootService {

    private final IRootFileService.Stub binder = new IRootFileService.Stub() {
        @Override
        public Bundle exists(String path) {
            try {
                return NativeFileBundles.successBoolean(NativeFileBridge.exists(path));
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle list(String path) {
            try {
                String[] entries = NativeFileBridge.list(path);
                return NativeFileBundles.successStringList(
                    entries != null ? Arrays.asList(entries) : Collections.emptyList()
                );
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle readBytes(String path) {
            try {
                return NativeFileBundles.successBytes(NativeFileBridge.readBytes(path));
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle readText(String path) {
            try {
                byte[] bytes = NativeFileBridge.readBytes(path);
                return NativeFileBundles.successString(new String(bytes, StandardCharsets.UTF_8));
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle writeBytes(String path, byte[] bytes) {
            try {
                NativeFileBridge.writeBytes(path, bytes);
                return NativeFileBundles.successUnit();
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle writeText(String path, String text) {
            try {
                NativeFileBridge.writeBytes(path, text.getBytes(StandardCharsets.UTF_8));
                return NativeFileBundles.successUnit();
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle delete(String path, boolean recursive) {
            try {
                NativeFileBridge.delete(path, recursive);
                return NativeFileBundles.successUnit();
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle mkdirs(String path) {
            try {
                NativeFileBridge.mkdirs(path);
                return NativeFileBundles.successUnit();
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle copy(String sourcePath, String targetPath, boolean overwrite) {
            try {
                NativeFileBridge.copy(sourcePath, targetPath, overwrite);
                return NativeFileBundles.successUnit();
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle move(String sourcePath, String targetPath, boolean overwrite) {
            try {
                NativeFileBridge.move(sourcePath, targetPath, overwrite);
                return NativeFileBundles.successUnit();
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }

        @Override
        public Bundle chmod(String path, String mode) {
            try {
                NativeFileBridge.chmod(path, mode);
                return NativeFileBundles.successUnit();
            } catch (Throwable throwable) {
                return NativeFileBundles.failure(throwable);
            }
        }
    };

    @Override
    public IBinder onBind(Intent intent) {
        return binder;
    }
}

File: androidSharedNative/src/main/kotlin/io/github/lumkit/tweak/sharednative/RootFileService.kt
package io.github.lumkit.tweak.sharednative

import android.content.Intent
import android.os.Bundle
import android.os.IBinder
import android.os.ParcelFileDescriptor
import com.topjohnwu.superuser.ipc.RootService

class RootFileService : RootService() {

    private val binder = object : IRootFileService.Stub() {
        override fun exists(path: String): Bundle {
            return FileServiceDelegate.exists(path)
        }

        override fun list(path: String): Bundle {
            return FileServiceDelegate.list(path)
        }

        override fun listEntries(path: String): Bundle {
            return FileServiceDelegate.listEntries(path)
        }

        override fun zipEntries(path: String): Bundle {
            return FileServiceDelegate.zipEntries(path)
        }

        override fun openReadOnlyFd(path: String): ParcelFileDescriptor {
            return FileServiceDelegate.openReadOnlyFd(path)
        }

        override fun openWriteOnlyFd(path: String, create: Boolean, truncate: Boolean): ParcelFileDescriptor {
            return FileServiceDelegate.openWriteOnlyFd(path, create, truncate)
        }

        override fun readBytes(path: String): Bundle {
            return FileServiceDelegate.readBytes(path)
        }

        override fun readText(path: String): Bundle {
            return FileServiceDelegate.readText(path)
        }

        override fun writeBytes(path: String, bytes: ByteArray): Bundle {
            return FileServiceDelegate.writeBytes(path, bytes)
        }

        override fun writeText(path: String, text: String): Bundle {
            return FileServiceDelegate.writeText(path, text)
        }

        override fun delete(path: String, recursive: Boolean): Bundle {
            return FileServiceDelegate.delete(path, recursive)
        }

        override fun mkdirs(path: String): Bundle {
            return FileServiceDelegate.mkdirs(path)
        }

        override fun copy(sourcePath: String, targetPath: String, overwrite: Boolean): Bundle {
            return FileServiceDelegate.copy(sourcePath, targetPath, overwrite)
        }

        override fun move(sourcePath: String, targetPath: String, overwrite: Boolean): Bundle {
            return FileServiceDelegate.move(sourcePath, targetPath, overwrite)
        }

        override fun chmod(path: String, mode: String): Bundle {
            return FileServiceDelegate.chmod(path, mode)
        }

        override fun readCpuCycles(coreIndex: Int): Bundle {
            return FileServiceDelegate.readCpuCycles(coreIndex)
        }

        override fun listInstalledApps(): Bundle {
            return FileServiceDelegate.listInstalledApps()
        }

        override fun getInstalledApp(packageName: String): Bundle {
            return FileServiceDelegate.getInstalledApp(packageName)
        }

        override fun unzipToDir(pfd: ParcelFileDescriptor, targetDir: String): Bundle {
            return FileServiceDelegate.unzipToDir(pfd, targetDir)
        }

        override fun unzipPathToDir(sourcePath: String, targetDir: String): Bundle {
            return FileServiceDelegate.unzipPathToDir(sourcePath, targetDir)
        }
    }

    override fun onBind(intent: Intent): IBinder {
        return binder
    }
}
